diff --git a/target/ppc/int_helper.c b/target/ppc/int_helper.c
--- a/target/ppc/int_helper.c
+++ b/target/ppc/int_helper.c
@@ -720,7 +720,7 @@
         bool prod_zero = !(a[i] & 0x7fffffff) || !(c[i] & 0x7fffffff);
 
         ok &= vfp_zon(a[i]) & vfp_zon(b[i]) & vfp_zon(c[i]);
-        ok &= prod_zero || is_inf || !(fabsf(fr) <= FLT_MIN);
+        ok &= (prod_zero && !nres) || is_inf || !(fabsf(fr) <= FLT_MIN);
         inf |= is_inf && !prod_zero;
         res[i] = ur ^ nres;
     }
