diff --git a/hw/display/Kconfig b/hw/display/Kconfig
index b3593fe..c170833 100644
--- a/hw/display/Kconfig
+++ b/hw/display/Kconfig
@@ -153,3 +153,8 @@ config MAC_PVG_PCI
     bool
     depends on MAC_PVG && PCI
     default y if PCI_DEVICES
+
+config QFB_PCI
+    bool
+    default y if PCI_DEVICES
+    depends on PCI
diff --git a/hw/display/meson.build b/hw/display/meson.build
index ffecedb..51456a5 100644
--- a/hw/display/meson.build
+++ b/hw/display/meson.build
@@ -140,6 +140,7 @@ if config_all_devices.has_key('CONFIG_VIRTIO_VGA')
   endif
 endif
 
+system_ss.add(when: 'CONFIG_QFB_PCI', if_true: files('qfb-pci.c'))
 system_ss.add(when: 'CONFIG_OMAP', if_true: files('omap_lcdc.c'))
 
 modules += { 'hw-display': hw_display_modules }
