diff --git a/hw/display/Kconfig b/hw/display/Kconfig
index 2250c74..1b02fa0 100644
--- a/hw/display/Kconfig
+++ b/hw/display/Kconfig
@@ -140,3 +140,8 @@ config XLNX_DISPLAYPORT
 
 config DM163
     bool
+
+config QFB_PCI
+    bool
+    default y if PCI_DEVICES
+    depends on PCI
diff --git a/hw/display/meson.build b/hw/display/meson.build
index 20a9497..7235958 100644
--- a/hw/display/meson.build
+++ b/hw/display/meson.build
@@ -140,6 +140,7 @@ if config_all_devices.has_key('CONFIG_VIRTIO_VGA')
   endif
 endif
 
+system_ss.add(when: 'CONFIG_QFB_PCI', if_true: files('qfb-pci.c'))
 system_ss.add(when: 'CONFIG_OMAP', if_true: files('omap_lcdc.c'))
 
 modules += { 'hw-display': hw_display_modules }
