diff --git a/home/hsp/src/openbios-master-clean/arch/ppc/qemu/init.c b/home/hsp/src/openbios-master-smp/arch/ppc/qemu/init.c
index 253394c..55218d7 100644
--- a/home/hsp/src/openbios-master-clean/arch/ppc/qemu/init.c
+++ b/home/hsp/src/openbios-master-smp/arch/ppc/qemu/init.c
@@ -350,6 +350,7 @@ cpu_add_pir_property(void)
     asm("mfspr %0, 1023\n"
         : "=r"(pir) :);
     PUSH(pir);
+    printk("Pir = %lu \n", pir);
     fword("encode-int");
     push_str("reg");
     fword("property");
@@ -381,9 +382,9 @@ static void
 cpu_g4_init(const struct cpudef *cpu)
 {
     cpu_generic_init(cpu);
-    cpu_add_pir_property();
+    //cpu_add_pir_property();
 
-    fword("finish-device");
+    //fword("finish-device");
 }
 
 #ifdef CONFIG_PPC_64BITSUPPORT
@@ -566,7 +567,7 @@ static const struct cpudef ppc_defs[] = {
         .dcache_block_size = 0x20,
         .tlb_sets = 0x40,
         .tlb_size = 0x80,
-        .initfn = cpu_750_init,
+        .initfn = cpu_g4_init,
     },
     {
         .iu_version = 0x0000c0000,
@@ -1073,11 +1074,24 @@ arch_of_init(void)
     push_str("reg");
     fword("property");
 
+for (int i= 0; i<temp; i++) {
     cpu = id_cpu();
     cpu->initfn(cpu);
-    printk("CPU type %s\n", cpu->name);
 
+    printk("CPU type %s\n", cpu->name);
     snprintf(buf, sizeof(buf), "/cpus/%s", cpu->name);
+    PUSH(i);
+    fword("encode-int");
+    push_str("reg");
+    fword("property");
+    if (i!=0){
+        push_str("stopped");
+        fword("encode-string");
+        push_str("state");
+        fword("property");
+        }
+    fword("finish-device");
+}
     ofmem_register(find_dev("/memory"), find_dev(buf));
     node_methods_init(buf);
 
